src/reconcile/rec_child.c

```diff
@@ -15,90 +15,80 @@
+     * page.
      *
-     * A visible update to be in READY state (i.e. not in LOCKED or PREPARED state), for truly
-     * visible to others.
+     * We expect the page to be clean after reconciliation. If there are invisible updates, abort
+     * eviction.
      */
-    if (F_ISSET(r, WT_REC_CLEAN_AFTER_REC | WT_REC_VISIBILITY_ERR) && page_del != NULL &&
-      __wt_page_del_active(session, ref, !F_ISSET(txn, WT_TXN_HAS_SNAPSHOT))) {
+    if (__wt_page_del_active(session, ref, !F_ISSET(txn, WT_TXN_HAS_SNAPSHOT))) {
         if (F_ISSET(r, WT_REC_VISIBILITY_ERR))
             WT_RET_PANIC(session, EINVAL, "reconciliation illegally skipped an update");
-        return (__wt_set_return(session, EBUSY));
+        if (F_ISSET(r, WT_REC_CLEAN_AFTER_REC))
+            return (__wt_set_return(session, EBUSY));
+        *statep = WT_CHILD_ORIGINAL;
+        return (0);
     }
 
     /*
-     * Deal with any underlying disk blocks.
-     *
-     * First, check to see if there is an address associated with this leaf: if there isn't, we're
-     * done, the underlying page is already gone. If the page still exists, check for any
-     * transactions in the system that might want to see the page's state before it's deleted.
-     *
-     * If any such transactions exist, we cannot discard the underlying leaf page to the block
-     * manager because the transaction may eventually read it. However, this write might be part of
-     * a checkpoint, and should we recover to that checkpoint, we'll need to delete the leaf page,
-     * else we'd leak it. The solution is to write a proxy cell on the internal page ensuring the
-     * leaf page is eventually discarded.
+     * A visible entry can be in a prepared state and checkpoints skip in-progress prepared changes.
+     * We can't race here, the entry won't be visible to the checkpoint, or will be in a prepared
+     * state, one or the other.
      *
-     * If no such transactions exist, we can discard the leaf page to the block manager and no cell
-     * needs to be written at all. We do this outside of the underlying tracking routines because
-     * this action is permanent and irrevocable. (Clearing the address means we've lost track of the
-     * disk address in a permanent way. This is safe because there's no path to reading the leaf
-     * page again: if there's ever a read into this part of the name space again, the cache read
-     * function instantiates an entirely new page.)
+     * We should never see an in-progress prepare in eviction: when we check to see if an internal
+     * page can be evicted, we check for an unresolved fast-truncate, which includes a fast-truncate
+     * in a prepared state, so it's an error to see that during eviction.
      */
-    if (ref->addr != NULL && !__wt_page_del_active(session, ref, true)) {
-        WT_RET(__wt_ref_block_free(session, ref));
+    WT_ORDERED_READ(prepare_state, page_del->prepare_state);
```

Comment:
Should we also add comment here about checkpoint?

Reply:
The comment immediately above it mentions checkpoint, I think?

Reply:
Oh, right.